### PR TITLE
889 HTTP logging

### PR DESCRIPTION
**Pull request by nimshi89**
### What

Added an optional response argument to the logging functions.
Link to list of expected values: https://github.com/ONSdigital/dp-standards/blob/main/LOGGING_STANDARDS.md
Added functions to support the logging of responses to the utility.py
Made the 'testing.py' so the method can be tested.

### How to review

run a simple script with logging and add a response as an argument

### Who can review

anyone

### CODE REVIEW

File: dpytools/logging/utility.py [hunk 1]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+    
+    end_date = strp_time + td
+
+    duration = end_date - strp_time
+    
+    return duration.microseconds * 1000 
+
+def get_content_length(res: Response)->int:
+    """
+    This funciton will try to get the 'Content-Lenght'
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "function" and "Content-Length"

File: dpytools/logging/utility.py [hunk 2]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+
+    strp_time = datetime.strptime(date, '%a, %d %b %Y %H:%M:%S GMT')
+    td = timedelta(microseconds=time_delta.microseconds)
+    
+    end_date = strp_time + td
+
+    return end_date.isoformat() + "Z"
+
+def calculate_duration_in_nanoseconds(time_delta: timedelta, date: str)->int:
+    """This funciton will convert the duration from Miliseconds to Nanoseconds."""
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "function"

File: dpytools/logging/utility.py [hunk 3]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+        return parsed_url.port
+
+def get_start_date(date: str) ->str:
+    
+    strp_time = datetime.strptime(date, '%a, %d %b %Y %H:%M:%S GMT')
+    
+    return strp_time.isoformat() + "Z"
+
+def get_end_date(time_delta: timedelta, date: str) ->str:
+    """This function will calculate the end_date by addint the duration to the start date."""
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "adding"

File: dpytools/logging/utility.py [hunk 4]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+
+def get_scheme(url: str) -> str:
+    """This function will return the scheme from the provided url."""
+
+    index = url.find('/')
+    scheme = url[0:index-1]
+    return scheme
+
+def get_domain(url: str) -> str:
+    """This funciton will return the domain name from the provided url."""
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "function"

File: dpytools/logging/utility.py [hunk 5]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+    #Getting the domain name
+    domain = parsed_url.netloc.split(':')[0]
+    return domain
+
+def get_port(url: str) ->int:
+    """This funciton will return the port number form the provided url."""
+    #Parsing url
+    parsed_url = urlparse(url)
+
+    #checking if the port was give if not scheking scheme for port number
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "checking scheme..."

File: dpytools/logging/utility.py [hunk 6]
```diff
@@ -50,3 +54,72 @@ def create_error_dict(error: Exception) -> List[Dict]:
...
+    """This funciton will return the domain name from the provided url."""
+    #Parsing url to extract the domain name
+    parsed_url = urlparse(url)
+
+    #Getting the domain name
+    domain = parsed_url.netloc.split(':')[0]
+    return domain
+
+def get_port(url: str) ->int:
+    """This funciton will return the port number form the provided url."""
```
Comment (NickPapONS) [inline dpytools/logging/utility.py]: Should be "function"

File: dpytools/logging/logger.py [hunk 7]
```diff
@@ -86,17 +115,17 @@ def _log(
         if self.flush_stdout_after_log_entry is True:
             sys.stdout.flush()
 
-    def debug(self, event: str, raw: str = None, data: Dict = None):
+    def debug(self, event: str, raw: str = None, data: Dict = None, response: requests.Response = None):
```
Comment (NickPapONS) [inline dpytools/logging/logger.py:118]: In the log_event dictionary above, the r_dict object in line 108 is assigned to the key `response_dict`, but in these logging functions (debug, info, warning) the response details are written simply as `response`, could that potentially lead to any issues or discrepancies? Not saying it should change, just wondering if anything could happen as a result of the argument assignment not having the same exact names. if not then safely ignore this

**Review by NickPapONS — CHANGES_REQUESTED** [review]
A couple of questions and some minor word changes in the docstrings, apart from that looks good.

File: dpytools/logging/logger.py (hunk 7 above)
Comment (nimshi89) [reply dpytools/logging/logger.py:118]: I don't think it would lead to any issue but yeah if it is a concern I can change it.

**Review by NickPapONS — APPROVED** [review]
Looks good to me, approved 👍

**Pull request merged**